src/torch_grid_utils/fftfreq_grid.py

```diff
@@ -110,16 +147,82 @@ def _construct_fftfreq_grid_2d(
     freq_y = torch.fft.fftfreq(h, d=dh, device=device)
     freq_x = last_axis_frequency_func(w, d=dw, device=device)
     h, w = rfft_shape(image_shape) if rfft is True else image_shape
-    freq_yy = einops.repeat(freq_y, 'h -> h w', w=w)
-    freq_xx = einops.repeat(freq_x, 'w -> h w', h=h)
-    return einops.rearrange([freq_yy, freq_xx], 'freq h w -> h w freq')
+    freq_yy = einops.repeat(freq_y, "h -> h w", w=w)
+    freq_xx = einops.repeat(freq_x, "w -> h w", h=h)
+    return einops.rearrange([freq_yy, freq_xx], "freq h w -> h w freq")
+
+
+def _apply_transform_matrix_2d(
+    frequency_grid: torch.Tensor,
+    transform_matrix: torch.Tensor,
+    device: torch.device | None = None,
+) -> torch.Tensor:
+    """Apply a 2x2 transformation matrix to a 2D frequency grid.
+
+    The input transform_matrix is a real-space transformation matrix A.
+    In Fourier space, coordinates transform by the inverse transpose:
+    k' = (A^-1)^T @ k.
+
+    Parameters
+    ----------
+    frequency_grid: torch.Tensor
+        Frequency grid with shape `(h, w, 2)`.
+    transform_matrix: torch.Tensor
+        2x2 real-space transformation matrix A. The frequency-space
+        transformation (A^-1)^T is automatically computed.
+    device: torch.device | None
+        PyTorch device for the transformation matrix.
+
+    Returns
+    -------
+    transformed_grid: torch.Tensor
+        Transformed frequency grid with shape `(h, w, 2)`.
+    """
+    # Ensure matrix is on the correct device
+    if device is not None:
+        transform_matrix = transform_matrix.to(device)
+    elif frequency_grid.device is not None:
+        transform_matrix = transform_matrix.to(frequency_grid.device)
+
+    # Validate matrix shape
+    if transform_matrix.shape != (2, 2):
+        raise ValueError(
+            f"transform_matrix must be a 2x2 matrix, "
+            f"got shape {transform_matrix.shape}"
+        )
+
+    # Ensure matrix is float type for matrix operations
+    transform_matrix = transform_matrix.float()
+
+    # Convert real-space transformation to frequency-space transformation
+    # If A is the real-space transform, frequency space uses (A^-1)^T
+    freq_transform = torch.linalg.inv(transform_matrix).T
+
+    # Apply transformation:
+    # [freq_y_new, freq_x_new]^T = freq_transform @ [freq_y, freq_x]^T
+    # frequency_grid has shape (h, w, 2), we need to apply matrix
+    # multiplication to the last dimension. Reshape to (h*w, 2) for batch
+    # matrix multiplication
+    h, w = frequency_grid.shape[:2]
+    freq_flat = einops.rearrange(frequency_grid, "h w freq -> (h w) freq")
+
+    # Apply transformation: (h*w, 2) @ (2, 2)^T = (h*w, 2)
+    # Note: freq_transform @ freq_flat^T would give (2, h*w), so we transpose
+    transformed_flat = torch.matmul(freq_flat, freq_transform.T)
+
+    # Reshape back to (h, w, 2)
+    transformed_grid = einops.rearrange(
+        transformed_flat, "(h w) freq -> h w freq", h=h, w=w
+    )
```

Comment:
```suggestion
    # [freq_y_new, freq_x_new]^T = freq_transform @ [freq_y, freq_x]^T
    # Apply transformation: (2, 2) @ (h, w, 2, 1) = (h, w, 2, 1)
    frequency_grid = einops.rearrange(frequency_grid, "h w freq_yx -> h w freq_yx 1")
    frequency_grid = freq_transform @ frequency_grid  # (h, w, freq_yx, 1)
    frequency_grid = einops.rearrange(frequency_grid, "h w freq_yx 1 -> h w freq_yx")
```

nit: I find the above more readable for matrix multiplies (order of operands never feels obvious to me in calls to np.matmul/torch.matmul)